### PR TITLE
OSASINFRA-3061: OpenStack Machine: Add FailureDomain

### PR DESCRIPTION
**Pull request by pierreprinetti**
Add a FailureDomain field to the OpenStack Machine providerSpec.

The placement instructions in this new field are designed to be unambiguously traced back to the Failure domain that originated them.

The existing zone instructions are individually incompatible with their FailureDomain counterpart; Port instructions instead are merged together.

### CODE REVIEW

File: machine/v1alpha1/types_openstack.go [hunk 1]
```diff
@@ -100,6 +101,28 @@ type OpenstackProviderSpec struct {
 
 	// The subnet that a set of machines will get ingress/egress traffic from
 	PrimarySubnet string `json:"primarySubnet,omitempty"`
+
+	// FailureDomain bundles placement instructions.
+	FailureDomain FailureDomain `json:"failureDomain,omitempty"`
+}
+
+// FailureDomain defines a bundle of fault domains across different modules: compute, storage and network.
```
Comment (mdbooth) [inline machine/v1alpha1/types_openstack.go:109]: // FailureDomain defines a set of correlated fault domains across different OpenStack services: compute, storage, and network.

File: machine/v1alpha1/types_openstack.go [hunk 2]
```diff
@@ -100,6 +101,28 @@ type OpenstackProviderSpec struct {
 
 	// The subnet that a set of machines will get ingress/egress traffic from
 	PrimarySubnet string `json:"primarySubnet,omitempty"`
+
+	// FailureDomain bundles placement instructions.
```
Comment (mdbooth) [inline machine/v1alpha1/types_openstack.go:105]: // FailureDomain defines the failure domain this machine will be created in.

**Review by mdbooth — APPROVED** [review]
I've included some suggested docstring changes, but code-wise this is what we need.

**Comment by pierreprinetti** [comment]
@mdbooth made those two changes. Also rebased. Thank you!

File: machine/v1alpha1/types_openstack.go [hunk 3]
```diff
@@ -65,6 +65,7 @@ type OpenstackProviderSpec struct {
 	FloatingIP string `json:"floatingIP,omitempty"`
 
 	// The availability zone from which to launch the server.
+	// It is an error to set both this and `ComputeAvailabilityZone` in `FailureDomain`.
```
Comment (JoelSpeed) [inline machine/v1alpha1/types_openstack.go:68]: Can we add a CEL validation that prevents this? (Edit: No, this is unstructured)

File: machine/v1alpha1/types_openstack.go [hunk 4]
```diff
@@ -100,6 +101,30 @@ type OpenstackProviderSpec struct {
 
 	// The subnet that a set of machines will get ingress/egress traffic from
 	PrimarySubnet string `json:"primarySubnet,omitempty"`
+
+	// FailureDomain defines the failure domain this machine will be
+	// created in.
+	FailureDomain FailureDomain `json:"failureDomain,omitempty"`
```
Comment (JoelSpeed) [inline machine/v1alpha1/types_openstack.go:107]: What's the motivation behind `omitempty` on all of these new fields here?

File: machine/v1alpha1/types_openstack.go [hunk 5]
```diff
@@ -100,6 +101,30 @@ type OpenstackProviderSpec struct {
...
+}
+
+// FailureDomain defines a set of correlated fault domains across different
+// OpenStack services: compute, storage, and network.
+type FailureDomain struct {
+	// ComputeAvailabilityZone is the name of a valid nova availability
+	// zone. The server will be created in this availability zone.
```
Comment (JoelSpeed) [inline machine/v1alpha1/types_openstack.go:114]: Should add the mutual exclusivity note here too

File: machine/v1alpha1/types_openstack.go [hunk 6]
```diff
@@ -100,6 +101,30 @@ type OpenstackProviderSpec struct {
...
+	// +optional
+	ComputeAvailabilityZone string `json:"computeAvailabilityZone,omitempty"`
+
+	// StorageAvailabilityZone is the name of a valid cinder availability
+	// zone. This will be the availability zone of the root volume if one is
+	// specified.
```
Comment (JoelSpeed) [inline machine/v1alpha1/types_openstack.go:120]: Should add mutual exclusivity stuff here too

File: machine/v1alpha1/types_openstack.go [hunk 7]
```diff
@@ -100,6 +101,30 @@ type OpenstackProviderSpec struct {
 
 	// The subnet that a set of machines will get ingress/egress traffic from
 	PrimarySubnet string `json:"primarySubnet,omitempty"`
+
+	// FailureDomain defines the failure domain this machine will be
```
Comment (JoelSpeed) [inline machine/v1alpha1/types_openstack.go:105]: Nit, first word should be the json tag
```suggestion
	// failureDomain defines the failure domain this machine will be
```

File: machine/v1alpha1/types_openstack.go [hunk 8]
```diff
@@ -356,6 +381,7 @@ type RootVolume struct {
 	// diskSize specifies the size, in GB, of the created root volume.
 	Size int `json:"diskSize,omitempty"`
 	// availabilityZone specifies the Cinder availability where the root volume will be created.
+	// It is an error to set both this and `StorageAvailabilityZone` in `FailureDomain`.
```
Comment (JoelSpeed) [inline machine/v1alpha1/types_openstack.go:384]: Use the json names in this text as that's what a user will expect to see, they won't know the Go names

**Comment by pierreprinetti** [comment]
The failure domain definition will only merge to openshift/api when it's GA, which is not in OCP v4.13.

**Comment by EmilienM** [comment]
I just rebased it so I can use it for initial prototype on CPMS.

**Comment by pierreprinetti** [comment]
@EmilienM Feel free to use or close this PR as needed.

**Pull request closed without merging**